### ISSUE TITLE
The problem with VisualShader/VisualScript nodes count

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#55267]
### Godot version

3.4 / 4.0

### System information

W10

### Issue description

Soo this has bothered me for a while, but we have a huge number of exposed classes for VisualShader/VisualScript. Each visual node has its dedicated resource type, which is overkill.

My main problem with this is that each of these nodes can be documented, but no one really bothers to do it and it impacts doc completion. I once made a PR that documented "all" VisualShader nodes, but before it was merged even more nodes were already added.

With VisualScript it's a bit better, because the nodes are edited in the inspector and you can actually see their documentation. But for VisualShader it's not the case; the docs aren't displayed anywhere and also VisualShader editor has its own description for every function. So documenting everything is pointless, but then we end up with empty docs.

Assuming this is a concern, is it maybe possible to unexpose the nodes? They could be internal classes and the actual node (the one saved in the scene) would only hold information about what type it is and and the property list would depend on that. This way we could have single class for VisualScript/VisualShader, so the class ref isn't so cluttered.

### Steps to reproduce

Just look at this
![4mojLMKP2G](https://user-images.githubusercontent.com/2223172/143062922-e069cfcb-4d25-4da6-8892-bcf09e25bcae.gif)

**Comment by Chaosus** [comment]
> My main problem with this is that each of these nodes can be documented, but no one really bothers to do it and it impacts doc completion. I once made a PR that documented "all" VisualShader nodes, but before it was merged even more nodes were already added.

A lot of these classes have already been documented, if there something new was added I don't see a problem documenting it at any time.

> Assuming this is a concern, is it maybe possible to unexpose the nodes?

Than users lose the ability to create a visual shader/script from code (I doubt that anyone using this, tho).

> But for VisualShader it's not the case; the docs aren't displayed anywhere and also VisualShader editor has its own description for every function.

Maybe better to fix it instead?

**Comment by akien-mga** [comment]
>> Assuming this is a concern, is it maybe possible to unexpose the nodes?
>
> Than users lose the ability to create a visual shader/script from code (I doubt that anyone using this, tho).

I think the main use case would be for editor plugins or tool scripts, which allow expanding the existing features of the shader and script graphs. I haven't seen such a plugin yet, but I can see how it would be useful and it would then need to have access to the public methods and properties of each node type if it has to place them in the graph.

>> But for VisualShader it's not the case; the docs aren't displayed anywhere and also VisualShader editor has its own description for every function.
>
> Maybe better to fix it instead?

Yeah this could be improved. I guess at least node descriptions could be reused, and maybe also property descriptions as tooltips?

**Comment by KoBeWi** [comment]
> Th**e**n users lose the ability to create a visual shader/script from code (I doubt that anyone using this, tho).

Not really. We could expose some ways to create and interact with the nodes; it would be less convenient than now though (no typing etc.)

> Maybe we could figure out a way to expose such APIs as "advanced" and hidden by default. They're still public but not shown in the editor help unless you toggle an advanced option. That could apply to any exposed node or resource that isn't typically used in scripts, but could be for advanced use cases (e.g. VisualShaderNode*, VisualScript*, RD*, SkeletonModification*, AnimationNode*).

That would be nice actually.

> I guess at least node descriptions could be reused, and maybe also property descriptions as tooltips?

Maybe right-clicking a node could open a context menu, with an option to open documentation for that node? Right now you don't even know what type of node it is.

**Comment by Gallilus** [comment]
Check out https://github.com/godotengine/godot-proposals/issues/3528.
If the family gives me some time I hope to have an alfa ready by the 31st
